src/systems/filecoin_mining/storage_mining/storage_miner_actor.go

```diff
@@ -3,19 +3,104 @@ package storage_mining
 import block "github.com/filecoin-project/specs/systems/filecoin_blockchain/struct/block"
 import poster "github.com/filecoin-project/specs/systems/filecoin_mining/storage_proving/poster"
+import power "github.com/filecoin-project/specs/systems/filecoin_blockchain/storage_power_consensus"
+import util "github.com/filecoin-project/specs/util"
+
+func (st *SectorTable_I) ActivePower() block.StoragePower {
+	return block.StoragePower(st.ActiveSectors_ * util.UVarint(st.SectorSize_))
+}
+
+func (st *SectorTable_I) InactivePower() block.StoragePower {
+	return block.StoragePower((st.CommittedSectors_ + st.RecoveringSectors_ + st.FailingSectors_) * util.UVarint(st.SectorSize_))
+}
+
+// called by CronActor to notify StorageMiner of PoSt Challenge
+func (sm *StorageMinerActor_I) NotifyOfPoStChallenge() {
+	if sm.isChallenged() {
+		// TODO: proper throw
+		panic("should not be challenged when isChallenged")
+	}
+
+	sm.isChallenged_ = true
+
+	// TODO: commit state change
+}
+
+// construct FaultReport
+// reset NewTerminatedFaults
+func (sm *StorageMinerActor_I) submitFaultReport(
+	newDeclaredFaults util.UVarint,
+	newDetectedFaults util.UVarint,
+	newTerminatedFaults util.UVarint,
+) {
+	faultReport := &power.FaultReport_I{
+		NewDeclaredFaults_:   newDeclaredFaults,
+		NewDetectedFaults_:   newDetectedFaults,
+		NewTerminatedFaults_: newTerminatedFaults,
+	}
+
+	sm.SectorTable().Impl().TerminatedSectors_ = util.UVarint(0)
+
+	panic(faultReport)
+	// TODO: Send(SPA, ProcessFaultReport(faultReport))
+}
+
+// construct PowerReport from SectorTable
+func (sm *StorageMinerActor_I) submitPowerReport() {
+	powerReport := &power.PowerReport_I{
+		ActivePower_:   sm.SectorTable().ActivePower(),
+		InactivePower_: sm.SectorTable().InactivePower(),
+	}
+
+	panic(powerReport)
+	// TODO: Send(SPA, ProcessPowerReport(powerReport))
+}
+
+func (sm *StorageMinerActor_I) onMissedPoSt() {
+
+	failingSectorNumbers := getSectorNums(sm.Sectors())
+	for _, sectorNo := range failingSectorNumbers {
+		sm.failSector(sectorNo, true)
+	}
+
+	sm.expireSectors()
+
+	// Note: NewDetectedFaults is now the sum of all
+	// previously active, committed, and recovering sectors minus expired ones
+	// and any previously Failing sectors that did not exceed MaxFaultCount
+	// Note: previously declared faults is now treated as part of detected faults
+	sm.submitFaultReport(
+		util.UVarint(0),                      // NewDeclaredFaults
+		sm.SectorTable().FailingSectors(),    // NewDetectedFaults
+		sm.SectorTable().TerminatedSectors(), // NewTerminatedFaults
+	)
+
+	sm.submitPowerReport()
+
+	// end of challenge
+	sm.isChallenged_ = false
+
+	// TODO: commit state change
+
+}
 
 // If a Post is missed (either due to faults being not declared on time or
 // because the miner run out of time, every sector is reported as failing
 // for the current proving period.
 func (sm *StorageMinerActor_I) CheckPoStSubmissionHappened() {
+	if !sm.isChallenged() {
```

Comment:
```go
if !sm.inChallengePeriod() {
  return // probably got challenged twice
}
```
and

```go
func (sm *StorageMinerActor_I) inChallengePeriod() bool {
  sinceLastChallenge := runtime.CurrentEpoch() - sm.challengeEpoch_
  return sinceLastChallenge < EPOCHS_CHALLENGE_PERIOD
}
```

Reply:
If a miner gets challenged and submits a successful PoSt, they will be out of the ChallengePeriod and sm.inChallengePeriod() should return false. If a miner gets challenged and did not submit PoSt successfully, they should still be in the challenge period. Hence, correct, if !sm.inChallengePeriod(), we should return here 👍.

Reply:
However, just using EpochTime is not enough, need to mark miner successfully submitted a PoSt.